vm/src/interpreter.cpp

```diff
@@ -159,279 +221,332 @@ bool apply_comparison(Opcode code, double a, double b) {
         throw InvalidBytecodeError("Unknown opcode");
     }
 }
-void run(const ConstantPool &pool, const char *bytes, size_t size, strings_t &strings) {
-    std::unique_ptr<char[]> locals_buf = std::make_unique<char[]>(1024);
-    OperandStack stack = OperandStack(1024);
-
-    unsigned int ip = 0;
-    char *locals = locals_buf.get();
-    int local_frame = 0;
-    while (ip < size) {
+
+inline void push_function_invocation(constant_index callee_identifier_idx,
+                                     const runtime_state &state,
+                                     OperandStack &caller_operands,
+                                     CallStack &call_stack) {
+
+    const std::string *callee_identifier = &state.pool.get_string(callee_identifier_idx);
+    const function_definition &callee_def = state.functions.at(callee_identifier);
+
+    caller_operands.pop_bytes(callee_def.parameters_byte_count);
+
+    call_stack.push_frame(callee_def, callee_identifier);
+}
+
+/**
+ * Will run a frame until it returns or pushes a new method inside the call_stack
+ * @return true if this function returned because the current frame has ended, or false if it returned because it pushed a new frame
+ */
+bool run_frame(runtime_state &state, stack_frame &frame, CallStack &call_stack, const char *instructions, size_t instruction_count) {
+    const ConstantPool &pool = state.pool;
+
+    // the instruction pointer
+    size_t &ip = *frame.instruction_pointer;
+    OperandStack &operands = frame.operands;
```

Comment:
I have small doubts about calling it *operands* instead of *stack*. While it adds unnecessary diff, it also loses the fact that this is a stack.

Reply:
it is the stack for operands.

Reply:
Operand stacks is a term that the JVM uses, while a stack is a more universal term for something to work with. Instructions do *push* and *pop* from the stack, they do not *place* operands like we would *place* locals. On the other hand, this is true that there are more than one type of stack.